src/base/DeadCodeDetector.ml

```diff
@@ -531,6 +531,332 @@ module DeadCodeDetector (SR : Rep) (ER : Rep) = struct
+          -> (
+            match ty with
+            | SType.Address (ContrAddr m) ->
+                List.iter (SType.IdLoc_Comp.Map.keys m) ~f:(fun id ->
+                    let name = get_id id in
+                    if Set.mem fields name && (not @@ Set.mem used_fields name)
+                    then
+                      warn1
+                        ("Unused field in the contract address type: "
+                       ^ as_error_string id)
+                        warning_level_dead_code
+                        (SType.TIdentifier.get_rep id))
+            | _ -> ())
+        | _ -> ());
+    (* Collect fields of the [cmod] contract with contract address types used
+       in this component. *)
+    Map.keys used_addresses
+    |> List.fold_left ~init:used_contract_params ~f:(fun used_ids_map used_id ->
+           match Map.find used_ids_map used_id with
+           | Some used_fields ->
+               let data =
+                 Map.find_exn used_addresses used_id |> Set.union used_fields
+               in
+               Map.set used_ids_map ~key:used_id ~data
+           | None ->
+               let data = Map.find_exn used_addresses used_id in
+               Map.set used_ids_map ~key:used_id ~data)
+
+  (** Checks for unused fields in contract address types occurred in contract
+      parameters and parameters of contract's components. *)
+  let check_param_contract_address_types cmod =
+    let contract_params =
+      List.fold_left cmod.contr.cparams ~init:emp_idsmap ~f:(fun m (id, ty) ->
+          update_contract_params_map m id ty)
+    in
+    let used_contract_params =
+      List.fold_left cmod.contr.ccomps ~init:emp_idsmap ~f:(fun used c ->
+          check_contract_address_types_in_params used contract_params c)
+    in
+    (* Report unused contract parameters with contract address types. *)
+    Map.keys contract_params
+    |> List.iter ~f:(fun param ->
+           match Map.find used_contract_params param with
+           | Some used_fields ->
+               let all_fields = Map.find_exn contract_params param in
+               let unused_fields = Set.diff all_fields used_fields in
+               if not @@ Set.is_empty unused_fields then
+                 List.iter cmod.contr.cparams ~f:(fun (id, _ty) ->
+                     if SCIdentifier.Name.equal param (SCIdentifier.get_id id)
+                     then
+                       Set.iter unused_fields ~f:(fun f ->
+                           warn1
+                             ("Unused field in the contract address type: "
+                             ^ SCIdentifier.Name.as_string f)
+                             warning_level_dead_code
+                             (ER.get_loc (SCIdentifier.get_rep id))))
+           | None ->
+               (* All the fields of [param] are unused, so we don't report it.
+                  It is an unused contract parameter.*)
+               ())
```

Comment:
Sorry, which ones do you mean?
We need a different logic to handle contract parameters, component parameters and ADT fields.

Reply:
The functions are `check_contract_address_types_in_params` and `check_param_contract_address_types`.

Looking at the code again I can see what you mean, but I'm surprised the logic is that different. Anyway, not a big deal.

Reply:
Well, I created small functions to reuse similar logic used in both functions.
Yes, it has the similar patterns like these `fold`s, but I think, it will be less readable if we extract functions from them.